File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.ts
function maxProfit(prices: number[]): number {
    let buy: number = prices[0];
    let profit: number = 0;
    let result: number = 0;

    for (let i = 1; i < prices.length; i++) {
        buy = Math.min(buy, prices[i]);

        const prev: number = profit;
        profit = Math.max(profit, prices[i] - buy);

        if (profit !== prev && profit > 0) {
            result += profit;
            profit = 0;
            buy = prices[i];
        }
    }

    return result;
}

File: leetcode/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
	public int maxProfit(int[] prices) {
        
		int buy = prices[0];
    int profit = 0;
    int result = 0;
      
    for (int i = 1; i < prices.length; i ++) {
	    buy = Math.min(buy, prices[i]);
      int prev = profit;
      profit = Math.max(profit, prices[i] - buy);
      if (profit != prev && profit > 0) {
          result += profit;
          profit = 0;
          buy = prices[i];
			}
    }
    return result;  
  }
}